### PR TITLE
build updates for lighttpd

### PR DESCRIPTION
**Pull request by gstrauss**
build updates for lighttpd

Update configure.ac and various #ifdefs to enable WolfSSL to
build features for use by lighttpd.

Change signature of wolfSSL_GetVersion() to take const arg.
Pass (const WOLFSSL*) to wolfSSL_GetVersion() for use with
SSL_CTX_set_info_callback(), where OpenSSL callback takes (const SSL *)
Implemented `wolfSSL_dup_CA_list`

### CODE REVIEW

**Comment by kaleb-himes** [comment]
@gstrauss has been approved as a contributor. It is OK to test this PR.

**Comment by gstrauss** [comment]
rebased onto tip of wolfssl master branch

File: configure.ac
Comment (dgarske) [inline configure.ac:3528]: @gstrauss : Why is single threaded required? This breaks use of `--enable-crlmonitor`... which should get disabled if single threaded mode.

Comment (gstrauss) [reply configure.ac:3528]: lighttpd is single threaded, and so lighttpd does not require the locking overhead built into wolfSSL, even though the locking overhead should be minimal compared to the crypto overhead.

Reading `configure.ac`, I see that `--enable-crl-monitor` is disabled by default, and is implicitly enabled with `--enable-jni`.  lighttpd does not enable either.

Since lighttpd does not use the CRL Monitor, would you please explain your concern about specifying -DSINGLE_THREADED for lighttpd?

Maybe, separate from this pull request, https://www.wolfssl.com/docs/wolfssl-manual/ch2/ can be updated to note that `--enable-crl-monitor` requires that `-DSINGLE_THREADED` *must not* be used.  Right now, the doc says only
> crl-monitor - enabling this option adds the ability to have wolfSSL actively monitor a specific CRL (Certificate Revocation List) directory.

Tangentially related question: does lighttpd need to take any action to re-initialize wolfSSL after lighttpd forks?  By default, lighttpd does not fork, but lighttpd can be configured to run multiple workers, which run in independent forked subprocesses after wolfSSL is configured in the main process.

Comment (dgarske) [reply configure.ac]: @gstrauss , I can understand lighttpd not using multi-threading. wolfSSL also does not require multi-threading. The issue is with the `./configure --enable-all`, which turns on both CRL monitor and lighttpd.

I went ahead and pushed a fix to disable CRL monitor for single threaded and not enable lighty with `--enable-all`.

I will let you know if this PR has any other Jenkins testing or peer review issues.

Thanks for your work on this!

**Comment by gstrauss** [comment]
@dgarske instead of the patch you added, could we modify the lighty section to *not add* `-DSINGLE_THREADED` when `--enable-all` has been set? `-DSINGLE_THREADED` is not required for lighttpd, but rather a slight performance enhancement.  On the other hand, if someone builds wolfSSL `--enable-all`, I would like the resulting library to be usable by lighttpd, too, so I see no reason to omit lighttpd from `--enable-all`

**Comment by gstrauss** [comment]
```
--- a/configure.ac
+++ b/configure.ac
@@ -3525,8 +3525,11 @@ then
     # recommended
     AM_CFLAGS="$AM_CFLAGS -DOPENSSL_NO_SSL2 -DOPENSSL_NO_SSL3 -DOPENSSL_NO_COMP"
     AM_CFLAGS="$AM_CFLAGS -DNO_OLD_TLS"
-    AM_CFLAGS="$AM_CFLAGS -DSINGLE_THREADED"
-    ENABLED_SINGLETHREADED="yes"
+    if test "$ENABLED_ALL" = "no"
+    then
+        AM_CFLAGS="$AM_CFLAGS -DSINGLE_THREADED"
+        ENABLED_SINGLETHREADED="yes"
+    fi
     # w/ lighttpd 1.4.56 once WolfSSL is updated to expose non-filesystem funcs
     #AM_CFLAGS="$AM_CFLAGS -DNO_FILESYSTEM"
     #ENABLED_FILESYSTEM=no
```

**Comment by gstrauss** [comment]
Oh, I already see that some of what I just posted is part of your patch.

The other features that lighttpd suggests disabling by default (e.g. `-DNO_OLD_TLS`) should probably also be in the `"$ENABLED_ALL" = "no" block.

Please add `enable_lighty=yes` back to the `$ENABLED_ALL` block near the top of the file.

**Comment by gstrauss** [comment]
I pushed a change which does what I mentioned above.  Please let me know if you'd like me to modify the patch, or squash various commits.

**Comment by dgarske** [comment]
Hi @gstrauss ,

Getting these errors now:

```
src/.libs/src_libwolfssl_la-ssl.o: In function `wolfSSLv3_client_method_ex':
/var/lib/jenkins/workspace/PRB-known-configs-A/src/ssl.c:11447: undefined reference to `MakeSSLv3'
src/.libs/src_libwolfssl_la-ssl.o: In function `wolfSSLv3_server_method_ex':
/var/lib/jenkins/workspace/PRB-known-configs-A/src/ssl.c:11839: undefined reference to `MakeSSLv3'
/usr/bin/ld: src/.libs/libwolfssl.so.24.1.0: hidden symbol `MakeSSLv3' isn't defined
```

I've already fixed this one in PR: https://github.com/wolfSSL/wolfssl/pull/3182

Also seeing this one now:

```
src/ssl.c:42943:7: error: no previous prototype for ‘wolfSSL_X509_get_ex_data’ [-Werror=missing-prototypes]
 void *wolfSSL_X509_get_ex_data(X509 *x509, int idx)
       ^~~~~~~~~~~~~~~~~~~~~~~~
src/ssl.c:42957:5: error: no previous prototype for ‘wolfSSL_X509_set_ex_data’ [-Werror=missing-prototypes]
 int wolfSSL_X509_set_ex_data(X509 *x509, int idx, void *data)
     ^~~~~~~~~~~~~~~~~~~~~~~~
```

**Comment by gstrauss** [comment]
Works for me with `--enable-all`
Works for me with `--enable-lighty`
Can you give me a hint what the configuration is when it fails so that I can reproduce more easily?

**Comment by gstrauss** [comment]
I added a commit to expose prototypes for `wolfSSL_X509_get_ex_data` and `wolfSSL_X509_set_ex_data` with OPENSSL_EXTRA_X509_SMALL.  That may have accidentally gotten lost in a previous rebase.

**Comment by gstrauss** [comment]
Please let me know if you'd like me to modify the patch, or squash various commits.

**Comment by dgarske** [comment]
Hi @gstrauss , Once PR https://github.com/wolfSSL/wolfssl/pull/3182 is merged in you can rebase it. Then it should be good. Thanks again for your help on this.

**Comment by gstrauss** [comment]
#3182 was merged.  I have rebased this PR.  Please let me know if further changes are needed.

BTW, I hadn't missed the wolfssl/ssl.h ifdef change for `wolfSSL_X509_(get|set)_ex_data`, but had included it in #3115, intended to follow this pull request.  The ifdef update is now included here.

**Comment by dgarske** [comment]
Hi @gstrauss ,

Was getting error on `./configure --enable-lighty --enable-sslv3`. I pushed a fix for that. Then this should be good.

David Garske, wolfSSL

**Comment by gstrauss** [comment]
@dgarske I replaced that commit with a minor change to allow sslv3 and tlsv1.0 with lighttpd if those are explicitly enabled with configure command.

It's not that lighttpd doesn't support it, it's that I would rather people get more secure defaults.  If they really need sslv3 or tls1.0, they have to explicitly enable those.

**Comment by dgarske** [comment]
Hi @gstrauss ,

There is a configuration that fails our OCSP test with these changes. I am debugging the root cause and will let you know.

Thanks,
David Garske, wolfSSL

**Comment by dgarske** [comment]
Hi @gstrauss , 

Tracked it down to an issue using `--enable-ipv6` with OCSP stapling enabled. 

The fix is here:
https://github.com/wolfSSL/wolfssl/pull/3204/commits/6a984da53f4e5abb940cb4b329eb71d395785f9f#diff-5018f87b263374b87eabf227e8028a63R182

Basically adding a `-b` to the server call to bind to any instead of the IPV6 localhost.

It is probably easiest to wait until that PR #3204 gets merged.

Thanks,
David Garske, wolfSSL

**Comment by dgarske** [comment]
Hi @gstrauss ,

Thanks for rebasing after the PR #3204 merge.

The current error reports are:

```
./configure --disable-filesystem && make
src/ssl.c:6846:5: error: no previous prototype for ‘wolfSSL_CTX_check_private_key’ [-Werror=missing-prototypes]
 int wolfSSL_CTX_check_private_key(const WOLFSSL_CTX* ctx)
     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

Also looks like --enable-lighty is having issues with `scripts/ocsp-stapling2.test`.

```
./configure --enable-opensslextra --enable-openssh --enable-lighty --enable-stunnel && make check
FAIL: scripts/ocsp-stapling2.test
```

Thanks,
David Garske, wolfSSL

**Comment by gstrauss** [comment]
The current version of lighttpd (lighttpd 1.4.55) requires the OpenSSL filesystem APIs.
A future version of lighttpd (lighttpd 1.4.56) will not need the OpenSSL filesystem APIs.  See also #3115 

The current version of lighttpd (lighttpd 1.4.55) does not support OCSP stapling.
A future version of lighttpd (lighttpd 1.4.56) will support OCSP stapling, if properly configured in lighttpd.conf

**Comment by gstrauss** [comment]
```
./configure --disable-filesystem && make
src/ssl.c:6846:5: error: no previous prototype for ‘wolfSSL_CTX_check_private_key’ [-Werror=missing-prototypes]
 int wolfSSL_CTX_check_private_key(const WOLFSSL_CTX* ctx)
     ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
See #3245 fbe0e04388 Tesfa Mael committed on 18 Aug, which, BTW, was 6 weeks after I originally filed this PR.
It was merged 24 Aug by @dgarske.  Today is 2 Sep.  How did the wolfSSL team miss this failure of the wolfSSL test suite for over a week?

Why is `#if !defined(NO_FILESYSTEM)` part of the patch?  wolfSSL should be able to check that a private key matches without using the filesystem APIs.
```
#if !defined(NO_FILESYSTEM) && !defined(NO_CHECK_PRIVATE_KEY)
  WOLFSSL_API int  wolfSSL_CTX_check_private_key(const WOLFSSL_CTX*);
#endif /* !NO_CERTS */
```

24bfea1a by @dgarske added NO_CHECK_PRIVATE_KEY
How much code is being saved in exchange for this added complexity, added documentation, etc.?  250 lines of code in `wc_CheckPrivateKey()` and 50 lines of code for `wolfSSL_CTX_check_private_key()`

**Comment by gstrauss** [comment]
I do not see the failure you reported for `FAIL: scripts/ocsp-stapling2.test` when I test the branch for this PR on my x86_64 laptop.  What branch is being tested?

```
$ cat /etc/fedora-release 
Fedora release 32 (Thirty Two)
$ git status -sbuno
## lighttpd-updates
$ ./configure --enable-opensslextra --enable-openssh --enable-lighty --enable-stunnel && make check
[...]
PASS: scripts/external.test
PASS: scripts/resume.test
PASS: scripts/openssl.test
PASS: scripts/psk.test
PASS: scripts/crl-revoked.test
PASS: scripts/tls13.test
PASS: testsuite/testsuite.test
PASS: scripts/google.test
PASS: tests/unit.test
PASS: scripts/ocsp.test
PASS: scripts/ocsp-stapling.test
PASS: scripts/ocsp-stapling-with-ca-as-responder.test
PASS: scripts/ocsp-stapling2.test
============================================================================
Testsuite summary for wolfssl 4.5.0
============================================================================
# TOTAL: 13
# PASS:  13
# SKIP:  0
# XFAIL: 0
# FAIL:  0
# XPASS: 0
# ERROR: 0
============================================================================
```

**Comment by dgarske** [comment]
Hi @gstrauss , I pushed a fix for wolfSSL_CTX_check_private_key to your branch. I don't think there is an OCSP failure. Sometimes our testing rig gets overwhelmed with randomized port conflict.

**Comment by gstrauss** [comment]
Unrelated to this PR, #3245 was a five-line patch made by the wolfSSL team.  2 lines were removal of trailing whitespace (unrelated to the patch), 1 line was blank, and 1 line was removal of a comment after #endif.  The remaining one line of the patch was incorrect and broke the test suite.  I do not understand why you are fixing #3245 with a one-line patch added to this unrelated PR (#3098), but also do not care as long as progress is made.

Are there any further known issues with this PR?  Please try to keep other issues separate.

**Review by dgarske — APPROVED** [review]
Hi @gstrauss , thanks for the feedback. Your PR is in good shape. The issue now is with a failing OCSP test, which is not related to your PR. We are working on a fix for that separately. I am sorry about the delay in getting this merged. Your PR is approved, but we cannot merge until the CI tests have passed.

**Comment by douzzer** [comment]
Superseded by #3115, itself superseded by #3474, which includes all constituent commits.

**Pull request closed without merging**
